src/lock_free/mwcas_descriptor.cpp

```diff
@@ -0,0 +1,158 @@
+    -> MwCASDescriptor *
+{
+  /*auto *page = _gc->GetPageIfPossible<MwCASDescriptor>();
+  auto *desc = (page == nullptr) ? new MwCASDescriptor{} : static_cast<MwCASDescriptor *>(page);
+  desc->target_cnt_ = 0;*/
+  // GCを無視した一時的な実装
+  return new MwCASDescriptor{};
+}
+
+/*##############################################################################
+ * Utilities
+ *############################################################################*/
+auto
+MwCASDescriptor::MwCAS()  //
+    -> bool
+{
+  // set a memory fence
+  stat_.store(kUndecided, kRelease);
+  const auto succeeded = MwCASInternal();
+  return succeeded;
+}
+
+//                wordのビット割り当て（実装後に消す）
+// |     63     |       62-50       |      49-47     |        46-0       |
+// | MwCAS Flag | Reference Counter | Begin Position |Descriptor Address |
+
+auto
+MwCASDescriptor::MwCASInternal(  //
+    const size_t begin_pos       //
+    ) -> bool
+{
+  const auto desc_addr = std::bit_cast<uint64_t>(this) | kMwCASFlag;
+  auto cur_stat = stat_.load(kSeqCst);
+  if (cur_stat == kUndecided) {
+    auto stat = kSucceeded;
+    for (size_t i = begin_pos; i < target_cnt_; ++i) {
+      auto &target = targets_[i];
+      if (!EmbedDescriptor(desc_addr, i, target.cnt)) {
+        stat = kFailed;
+        break;
+      }
+    }
+    cur_stat = stat_.load(kSeqCst);
+    if (cur_stat == kUndecided && stat_.compare_exchange_strong(cur_stat, stat, kSeqCst, kSeqCst)) {
+      cur_stat = stat;
+    }
+  }
+
+  auto ret = (cur_stat == kSucceeded);
+  auto target_cnt_sum = 0;
+  if (ret) {
+    for (size_t i = 0; i < target_cnt_; ++i) {
+      auto &target = targets_[i];
+      auto expected = target.addr->load(kSeqCst);
+      if ((expected ^ desc_addr) & kAddrMask) continue;
+      target.addr->compare_exchange_strong(expected, target.new_val, target.fence, target.fence);
+      target_cnt_sum += target.cnt;
+    }
+  } else {
+    for (size_t i = 0; i < target_cnt_; ++i) {
+      auto &target = targets_[i];
+      auto expected = target.addr->load(kSeqCst);
+      if ((expected ^ desc_addr) & kAddrMask) continue;
+      target.addr->compare_exchange_strong(expected, target.old_val, target.fence, target.fence);
+      target_cnt_sum += target.cnt;
+    }
+  }
+
+  exit_cnt_++;
+  if (exit_cnt_ == target_cnt_sum) {
+    // GC処理（未実装）
+  }
+
+  return ret;
+}
+
+auto
+MwCASDescriptor::EmbedDescriptor(  //
+    const uint64_t desc_addr,
+    const size_t pos,
+    const size_t cnt)  //
+    -> bool
+{
+  auto &target = targets_[pos];
+  auto *addr = target.addr;
+  auto word = addr->load(kSeqCst);
+  while (true) {
+    while (word & kMwCASFlag) {
+      if (((word ^ desc_addr) & kAddrMask) == 0) return true;
+      const auto another_word = word;
+      std::this_thread::sleep_for(kBackOffTime);
+      word = addr->load(kSeqCst);
+      if (((word ^ another_word) | ~kCntMask) == 0)  // Equal except for cnt.
+        continue;                                    // other threads modified this field
```

Comment:
ここは直に比較すればOKです．対象アドレスはソースコードのコメント部分にあるように，

```cpp
 |     63     |       62-50       |      49-47     |        46-0       |
 | MwCAS Flag | Reference Counter | Begin Position |Descriptor Address |
```

という割当ですが，

1. MwCASフラグが違う ➔ old/new valueに変更済み ➔ 再試行
2. Reference Counterが違う ➔ 誰かが追実行中に参加 ➔ 再度バックオフ
3. Begin Positionが違う ➔ old/new valueに変更済み or 別の記述子が埋込み済み ➔ 再試行
4. Descriptor Addressが違う ➔ old/new valueに変更済み or 別の記述子が埋込み済み ➔ 再試行

なので，追実行中を開始ないといけないのは完全一致した場合のみです．

```suggestion
      if (word != another_word) continue;  // other threads modified this field
        
```